File: backend/app/schemas/practice.py
from datetime import datetime

from pydantic import BaseModel


class DailyPracticeGenerateRequest(BaseModel):
    subject: str = "chinese"


class PracticeQuestionItem(BaseModel):
    question: str
    options: list[str]
    correct_index: int
    explanation: str


class DailyPracticeResponse(BaseModel):
    id: int
    subject: str
    topic: str
    questions: list[PracticeQuestionItem]
    score: int | None = None
    total: int = 0
    created_at: datetime


class SubmitPracticeRequest(BaseModel):
    answers: list[int]


class PracticeHistoryItem(BaseModel):
    id: int
    subject: str
    topic: str
    score: int | None
    total: int
    created_at: datetime


class PracticeHistoryResponse(BaseModel):
    items: list[PracticeHistoryItem]
    streak: int

File: frontend/src/api/practice.ts
import { request } from './config'

export interface PracticeQuestion {
  question: string
  options: string[]
  correct_index: number
  explanation: string
}

export interface DailyPracticeResponse {
  id: number
  subject: string
  topic: string
  questions: PracticeQuestion[]
  score: number | null
  total: number
  created_at: string
}

export function generateDailyPractice(subject: string = 'chinese'): Promise<DailyPracticeResponse> {
  return request<DailyPracticeResponse>({
    url: '/practice/generate',
    method: 'POST',
    data: { subject },
  })
}

export function submitPractice(id: number, answers: number[]): Promise<DailyPracticeResponse> {
  return request<DailyPracticeResponse>({
    url: `/practice/${id}/submit`,
    method: 'POST',
    data: { answers },
  })
}

export interface PracticeHistoryItem {
  id: number
  subject: string
  topic: string
  score: number | null
  total: number
  created_at: string
}

export interface PracticeHistoryResponse {
  items: PracticeHistoryItem[]
  streak: number
}

export function getPracticeHistory(): Promise<PracticeHistoryResponse> {
  return request<PracticeHistoryResponse>({ url: '/practice/history', method: 'GET' })
}
